File: sdk/python/src/openai_codex_security/targets.py
from __future__ import annotations

import os
import subprocess
from collections.abc import Sequence
from dataclasses import dataclass
from pathlib import Path
from typing import Literal, TypeAlias

from .errors import InvalidTargetError

PathInput: TypeAlias = str | os.PathLike[str]
ScanMode: TypeAlias = Literal["standard", "deep"]


@dataclass(frozen=True, slots=True)
class DiffTarget:
    """A committed-ref or working-tree Git diff target."""

    kind: Literal["refs", "working_tree"]
    base: str
    head: str | None = None

    def __post_init__(self) -> None:
        if self.kind not in ("refs", "working_tree"):
            raise InvalidTargetError(f"Unsupported diff target kind: {self.kind}")
        if not isinstance(self.base, str) or not self.base:
            raise InvalidTargetError("The diff base ref must be non-empty.")
        if self.kind == "refs" and (not isinstance(self.head, str) or not self.head):
            raise InvalidTargetError("Git diff refs must include a non-empty head ref.")
        if self.kind == "working_tree" and self.head is not None:
            raise InvalidTargetError("Working-tree targets cannot specify a head ref.")

    @classmethod
    def refs(cls, *, base: str, head: str = "HEAD") -> DiffTarget:
        if not base or not head:
            raise InvalidTargetError("Git diff refs must be non-empty.")
        return cls(kind="refs", base=base, head=head)

    @classmethod
    def working_tree(
        cls,
        *,
        base: str = "HEAD",
    ) -> DiffTarget:
        if not base:
            raise InvalidTargetError("The working-tree base ref must be non-empty.")
        return cls(kind="working_tree", base=base)


ScanTarget: TypeAlias = Literal["repository"] | DiffTarget | Sequence[PathInput]


@dataclass(frozen=True, slots=True)
class NormalizedTarget:
    kind: Literal["repository", "paths", "refs", "working_tree"]
    paths: tuple[str, ...] = ()
    base: str | None = None
    head: str | None = None
    base_ref: str | None = None
    head_ref: str | None = None


def normalize_repository(repository: PathInput) -> Path:
    path = Path(repository).expanduser().resolve()
    if not path.is_dir():
        raise InvalidTargetError(f"Repository is not a directory: {path}")
    return path


def normalize_target(repository: Path, target: ScanTarget) -> NormalizedTarget:
    repository = normalize_repository(repository)
    if target == "repository":
        return NormalizedTarget(kind="repository")

    if isinstance(target, DiffTarget):
        _require_git_repository(repository)
        base = _resolve_git_ref(repository, target.base)
        if target.kind == "refs":
            if target.head is None:  # Defensive for dynamically typed callers.
                raise InvalidTargetError("Git diff refs must include a non-empty head ref.")
            head = _resolve_git_ref(repository, target.head)
            return NormalizedTarget(
                kind="refs",
                base=base,
                head=head,
                base_ref=target.base,
                head_ref=target.head,
            )
        if target.kind == "working_tree":
            head = _resolve_git_ref(repository, "HEAD")
            return NormalizedTarget(
                kind="working_tree",
                base=base,
                head=head,
                base_ref=target.base,
                head_ref="HEAD",
            )
        raise InvalidTargetError(f"Unsupported diff target kind: {target.kind}")

    if isinstance(target, str) or not isinstance(target, Sequence):
        raise InvalidTargetError(
            "Scan target must be 'repository', a DiffTarget, or a sequence of paths."
        )
    if not target:
        raise InvalidTargetError("A path scan target must contain at least one path.")

    paths: list[str] = []
    for value in target:
        if not isinstance(value, (str, os.PathLike)):
            raise InvalidTargetError(
                "Path scan targets must contain only strings or path-like values."
            )
        if os.fspath(value) == "":
            raise InvalidTargetError("Path scan targets must not contain an empty path.")
        path = Path(value).expanduser()
        absolute = path.resolve() if path.is_absolute() else (repository / path).resolve()
        try:
            relative = absolute.relative_to(repository)
        except ValueError as exc:
            raise InvalidTargetError(f"Path target is outside the repository: {value}") from exc
        if not absolute.exists():
            raise InvalidTargetError(f"Path target does not exist: {value}")
        normalized = relative.as_posix()
        if normalized not in paths:
            paths.append(normalized)
    return NormalizedTarget(kind="paths", paths=tuple(paths))


def _require_git_repository(repository: Path) -> None:
    result = subprocess.run(
        ["git", "rev-parse", "--show-toplevel"],
        cwd=repository,
        check=False,
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        raise InvalidTargetError(f"Diff targets require a Git repository: {repository}")
    root = Path(result.stdout.strip()).resolve()
    if root != repository:
        raise InvalidTargetError(f"Diff target repository must be the Git worktree root: {root}")


def _resolve_git_ref(repository: Path, ref: str) -> str:
    result = subprocess.run(
        ["git", "rev-parse", "--verify", "--end-of-options", f"{ref}^{{commit}}"],
        cwd=repository,
        check=False,
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        raise InvalidTargetError(f"unknown Git ref: {ref}")
    return result.stdout.strip()


def repository_revision(repository: Path) -> str | None:
    result = subprocess.run(
        ["git", "rev-parse", "--verify", "HEAD^{commit}"],
        cwd=repository,
        check=False,
        capture_output=True,
        text=True,
    )
    return result.stdout.strip() if result.returncode == 0 else None

File: sdk/typescript/src/targets.ts
import { execFile as execFileCallback } from "node:child_process";
import { existsSync } from "node:fs";
import { lstat, realpath, stat } from "node:fs/promises";
import { homedir } from "node:os";
import { isAbsolute, join, relative, resolve, sep } from "node:path";
import { promisify } from "node:util";
import { InvalidTargetError } from "./errors.js";

const execFile = promisify(execFileCallback);

export type ScanMode = "standard" | "deep";
export type DiffTargetKind = "refs" | "working_tree";

export interface DiffTargetOptions {
  kind: DiffTargetKind;
  base: string;
  head?: string;
}

export class DiffTarget {
  readonly #brand = "DiffTarget";
  public readonly kind: DiffTargetKind;
  public readonly base: string;
  public readonly head?: string;

  public constructor(options: DiffTargetOptions) {
    void this.#brand;
    this.kind = options.kind;
    this.base = options.base;
    this.head = options.head;
    if (this.kind !== "refs" && this.kind !== "working_tree") {
      throw new InvalidTargetError(
        `Unsupported diff target kind: ${String(this.kind)}`,
      );
    }
    if (typeof this.base !== "string" || this.base.length === 0) {
      throw new InvalidTargetError("The diff base ref must be non-empty.");
    }
    if (
      this.kind === "refs" &&
      (typeof this.head !== "string" || this.head.length === 0)
    ) {
      throw new InvalidTargetError(
        "Git diff refs must include a non-empty head ref.",
      );
    }
    if (this.kind === "working_tree" && this.head !== undefined) {
      throw new InvalidTargetError(
        "Working-tree targets cannot specify a head ref.",
      );
    }
    Object.freeze(this);
  }

  public static refs({
    base,
    head = "HEAD",
  }: {
    base: string;
    head?: string;
  }): DiffTarget {
    return new DiffTarget({ kind: "refs", base, head });
  }

  public static workingTree({
    base = "HEAD",
  }: { base?: string } = {}): DiffTarget {
    return new DiffTarget({ kind: "working_tree", base });
  }
}

export type ScanTarget = "repository" | DiffTarget | readonly string[];
export type NormalizedTargetKind =
  | "repository"
  | "paths"
  | "refs"
  | "working_tree";

export interface NormalizedTarget {
  kind: NormalizedTargetKind;
  paths: readonly string[];
  base?: string;
  head?: string;
  baseRef?: string;
  headRef?: string;
}

export async function normalizeRepository(
  repository: string,
  signal?: AbortSignal,
): Promise<string> {
  const candidate = resolveRepositoryPath(repository);
  let canonical: string;
  try {
    canonical = await abortable(() => realpath(candidate), signal);
    if (!(await abortable(() => stat(canonical), signal)).isDirectory()) {
      throw new Error("not a directory");
    }
  } catch (error) {
    throwIfAborted(signal);
    throw new InvalidTargetError(
      `Repository is not a directory: ${candidate}`,
      {
        cause: error,
      },
    );
  }
  return canonical;
}

export function resolveRepositoryPath(repository: string): string {
  return resolve(expandHome(repository));
}

export async function enclosingGitWorktreeRoot(
  repository: string,
  signal?: AbortSignal,
): Promise<string | null> {
  let current = repository;
  let root: string | null = null;
  while (true) {
    try {
      await abortable(() => lstat(join(current, ".git")), signal);
      root = current;
    } catch (error) {
      throwIfAborted(signal);
      if (
        typeof error !== "object" ||
        error === null ||
        !("code" in error) ||
        (error.code !== "ENOENT" && error.code !== "ENOTDIR")
      ) {
        throw new InvalidTargetError(
          `Unable to inspect the enclosing Git worktree: ${current}`,
          { cause: error },
        );
      }
    }
    const parent = resolve(current, "..");
    if (parent === current) return root;
    current = parent;
  }
}

export async function normalizeTarget(
  repository: string,
  target: ScanTarget,
  signal?: AbortSignal,
): Promise<NormalizedTarget> {
  const root = await normalizeRepository(repository, signal);
  throwIfAborted(signal);
  if (target === "repository") {
    return { kind: "repository", paths: [] };
  }

  if (target instanceof DiffTarget) {
    if (target.kind !== "refs" && target.kind !== "working_tree") {
      throw new InvalidTargetError(
        `Unsupported diff target kind: ${String(target.kind)}`,
      );
    }
    if (typeof target.base !== "string" || target.base.length === 0) {
      throw new InvalidTargetError("The diff base ref must be non-empty.");
    }
    if (
      target.kind === "refs" &&
      (typeof target.head !== "string" || target.head.length === 0)
    ) {
      throw new InvalidTargetError(
        "Git diff refs must include a non-empty head ref.",
      );
    }
    if (target.kind === "working_tree" && target.head !== undefined) {
      throw new InvalidTargetError(
        "Working-tree targets cannot specify a head ref.",
      );
    }
    await requireGitRepository(root, signal);
    const base = await resolveGitRef(root, target.base, signal);
    if (target.kind === "refs") {
      const head = target.head;
      if (typeof head !== "string" || head.length === 0) {
        throw new InvalidTargetError(
          "Git diff refs must include a non-empty head ref.",
        );
      }
      return {
        kind: "refs",
        paths: [],
        base,
        head: await resolveGitRef(root, head, signal),
        baseRef: target.base,
        headRef: head,
      };
    }
    return {
      kind: "working_tree",
      paths: [],
      base,
      head: await resolveGitRef(root, "HEAD", signal),
      baseRef: target.base,
      headRef: "HEAD",
    };
  }

  if (!Array.isArray(target)) {
    throw new InvalidTargetError(
      "Scan target must be 'repository', a DiffTarget, or an array of paths.",
    );
  }
  if (target.length === 0) {
    throw new InvalidTargetError(
      "A path scan target must contain at least one path.",
    );
  }

  const paths: string[] = [];
  for (const value of target) {
    throwIfAborted(signal);
    if (typeof value !== "string") {
      throw new InvalidTargetError(
        "Path scan targets must contain only strings.",
      );
    }
    if (value.length === 0) {
      throw new InvalidTargetError(
        "Path scan targets must not contain an empty path.",
      );
    }
    const candidate = isAbsolute(expandHome(value))
      ? resolve(expandHome(value))
      : resolve(root, expandHome(value));
    if (!existsSync(candidate)) {
      throw new InvalidTargetError(`Path target does not exist: ${value}`);
    }
    let canonical: string;
    try {
      canonical = await abortable(() => realpath(candidate), signal);
    } catch (error) {
      throwIfAborted(signal);
      throw new InvalidTargetError(`Path target does not exist: ${value}`, {
        cause: error,
      });
    }
    const relativePath = relative(root, canonical);
    if (
      relativePath === ".." ||
      relativePath.startsWith(`..${sep}`) ||
      isAbsolute(relativePath)
    ) {
      throw new InvalidTargetError(
        `Path target is outside the repository: ${value}`,
      );
    }
    const normalized = relativePath.split(sep).join("/") || ".";
    if (!paths.includes(normalized)) {
      paths.push(normalized);
    }
  }
  return { kind: "paths", paths };
}

export function validateMode(target: NormalizedTarget, mode: ScanMode): void {
  if (mode !== "standard" && mode !== "deep") {
    throw new InvalidTargetError(`Unsupported scan mode: ${String(mode)}`);
  }
  if (
    mode === "deep" &&
    (target.kind === "refs" || target.kind === "working_tree")
  ) {
    throw new InvalidTargetError(
      "Deep mode supports repository and path targets only.",
    );
  }
}

export async function repositoryRevision(
  repository: string,
  signal?: AbortSignal,
): Promise<string | null> {
  try {
    return await gitOutput(
      repository,
      ["rev-parse", "--verify", "HEAD^{commit}"],
      signal,
    );
  } catch {
    throwIfAborted(signal);
    return null;
  }
}

async function requireGitRepository(
  repository: string,
  signal?: AbortSignal,
): Promise<void> {
  let root: string;
  try {
    root = await gitOutput(
      repository,
      ["rev-parse", "--show-toplevel"],
      signal,
    );
  } catch (error) {
    throwIfAborted(signal);
    throw new InvalidTargetError(
      `Diff targets require a Git repository: ${repository}`,
      {
        cause: error,
      },
    );
  }
  const canonicalRoot = await abortable(() => realpath(root), signal);
  if (canonicalRoot !== repository) {
    throw new InvalidTargetError(
      `Diff target repository must be the Git worktree root: ${canonicalRoot}`,
    );
  }
}

async function resolveGitRef(
  repository: string,
  ref: string,
  signal?: AbortSignal,
): Promise<string> {
  try {
    return await gitOutput(
      repository,
      ["rev-parse", "--verify", "--end-of-options", `${ref}^{commit}`],
      signal,
    );
  } catch (error) {
    throwIfAborted(signal);
    throw new InvalidTargetError(`unknown Git ref: ${ref}`, { cause: error });
  }
}

async function gitOutput(
  repository: string,
  args: readonly string[],
  signal?: AbortSignal,
): Promise<string> {
  throwIfAborted(signal);
  const { stdout } = await execFile("git", args, {
    cwd: repository,
    encoding: "utf8",
    signal,
  });
  return stdout.trim();
}

async function abortable<T>(
  operation: () => Promise<T>,
  signal?: AbortSignal,
): Promise<T> {
  if (signal === undefined) return await operation();
  throwIfAborted(signal);
  return await new Promise<T>((resolvePromise, reject) => {
    const onAbort = (): void => reject(abortReason(signal));
    signal.addEventListener("abort", onAbort, { once: true });
    void operation().then(
      (value) => {
        signal.removeEventListener("abort", onAbort);
        resolvePromise(value);
      },
      (error: unknown) => {
        signal.removeEventListener("abort", onAbort);
        reject(error);
      },
    );
  });
}

function throwIfAborted(signal?: AbortSignal): void {
  if (signal?.aborted === true) throw abortReason(signal);
}

function abortReason(signal: AbortSignal): unknown {
  return (
    signal.reason ??
    new DOMException("The operation was aborted.", "AbortError")
  );
}

function expandHome(value: string): string {
  if (value === "~") {
    return homedir();
  }
  if (value.startsWith("~/") || value.startsWith("~\\")) {
    return resolve(homedir(), value.slice(2).replace(/^[/\\]+/, ""));
  }
  return value;
}
